### PR TITLE
Support VS System theme

### LINKED ISSUE #74416: Syntax highlight colors differ between Dark and System modes
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Syntax-highlight-colors-differ-between-D/10431662)._

---
Version 17.7.0 Preview 6.0

With Dark theme selected, methods are yellow, while with System setting they are white.

---
#### Feedback Bot on 8/6/2023, 09:15 PM: 

(private comment, text removed)
#### Feedback Bot on 8/10/2023, 02:31 AM: 

(private comment, text removed)
#### Feedback Bot on 8/16/2023, 03:24 AM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by JoeRobich**
The System theme has a unique ThemeId but reuses the editor colors from the Light and Dark themes. When the System theme is selected we will now check the registry to see whether Apps should be using the light theme and then apply the appropriate color scheme.

I also noticed while testing that VSColorTheme_ThemeChanged fires multiple times in a row when the Windows theme is changed.  We only need to apply the color scheme colors once in response. I have added a field to track whether we have an update pending.

Resolves #74416

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
> We only need to apply the color scheme colors once in response. I have added a field to track whether we have an update pending.

Another way to handle this, which may be cleaner, is to use an AsyncBatchingWorkQueue to queue up the work. That way if there is a flurry of notifications, it will only end up causing one update to happen.

**Comment by JoeRobich** [comment]
> Another way to handle this, which may be cleaner, is to use an AsyncBatchingWorkQueue to queue up the work.

@CyrusNajmabadi Updated to use a work queue. Thanks for the suggestion.

File: src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs [hunk 1]
```diff
@@ -56,6 +61,11 @@ internal sealed partial class ColorSchemeApplier
         _settings = new ColorSchemeSettings(threadingContext, _serviceProvider, globalOptions);
         _colorSchemes = ColorSchemeSettings.GetColorSchemes();
         _classificationVerifier = new ClassificationVerifier(threadingContext, fontAndColorStorage, _colorSchemes);
+        _workQueue = new(
+            TimeSpan.FromMilliseconds(1000),
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs:65]: try DelayTimeSpan.Idle

File: src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs [hunk 2]
```diff
@@ -56,6 +61,11 @@ internal sealed partial class ColorSchemeApplier
...
+            async (token) => await QueueColorSchemeUpdateAsync().ConfigureAwait(false),
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs:66]:
```suggestion
            async _ => await QueueColorSchemeUpdateAsync().ConfigureAwait(false),
```

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs:66]: or, just replace this entirely with `QueueColorSchemeUpdateAsync` as the entire argument.

File: src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs [hunk 3]
```diff
@@ -56,6 +61,11 @@ internal sealed partial class ColorSchemeApplier
...
+            listenerProvider.GetListener(FeatureAttribute.ColorScheme),
+            CancellationToken.None);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs:68]:
```suggestion
            threadingContext.DisposalToken);
```

File: src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs [hunk 4]
```diff
@@ -106,10 +116,13 @@ private async Task UpdateColorSchemeAsync(CancellationToken cancellationToken)
...
+        => _workQueue.AddWork();
 
     private Task ColorSchemeChangedAsync(object sender, PropertyChangedEventArgs args)
-        => QueueColorSchemeUpdateAsync();
+    {
+        _workQueue.AddWork();
+        return Task.CompletedTask;
+    }
 
     private async Task QueueColorSchemeUpdateAsync()
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs:127]:
```suggestion
    private async ValueTask QueueColorSchemeUpdateAsync(CancellationToken cancellationToken)
```

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs:127]: also, use thei cancellationToken in the StartOnIdle call below.

**Comment by CyrusNajmabadi** [comment]
LMK when you make the changes, and i'll quickly review.

**Comment by JoeRobich** [comment]
@CyrusNajmabadi That cleaned up nicely. Thanks!

File: src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs [hunk 5]
```diff
@@ -106,15 +116,18 @@ private async Task UpdateColorSchemeAsync(CancellationToken cancellationToken)
...
 
     private Task ColorSchemeChangedAsync(object sender, PropertyChangedEventArgs args)
-        => QueueColorSchemeUpdateAsync();
+    {
+        _workQueue.AddWork();
+        return Task.CompletedTask;
+    }
 
-    private async Task QueueColorSchemeUpdateAsync()
+    private async ValueTask QueueColorSchemeUpdateAsync(CancellationToken token)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs:127]: nit: 
```suggestion
    private async ValueTask QueueColorSchemeUpdateAsync(CancellationToken cancellationToken)
```

File: src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs (hunk 1 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/ColorSchemes/ColorSchemeApplier.cs]: 💭 It seems like `Idle` could leave a really strange visual state, where `Short` would be equally effective but with less transition time.

**Pull request merged**